src/graphics/env-lighting.js

```diff
@@ -0,0 +1,397 @@
+     * @param {string} [options.distribution] - The specular distribution, either 'phong' or 'ggx'. Default is 'ggx'.
+     * @returns {Texture} The prefiltered cubemap
+     */
+    static generateReflection(source, options) {
+        const device = source.device;
+        const format = lightingPixelFormat(device);
+        const result = createCubemap(device, 128, format);
+
+        // generate the top level specular - just a copy of source
+        reprojectTexture(source, result, {
+            numSamples: 1
+        });
+
+        // generate convolved mipmaps
+        const levels = calcLevels(result.width, result.height);
+        for (let i = 1; i < levels; ++i) {
+            const level = createCubemap(device, Math.max(1, result.width >> i), format);
+            reprojectTexture(source, level, {
+                numSamples: options?.numSamples || 1024,
+                distribution: options?.distribution || 'ggx',
+                specularPower: Math.max(1, 2048 >> (i * 2))
+            });
+            copyMipmap(result, level);
+            level.destroy();
+        }
+
+        // set filtering while bypassing engine
+        device.setTexture(result, 0);
+        device.gl.texParameteri(result._glTarget, device.gl.TEXTURE_MAG_FILTER, device.gl.LINEAR);
+        device.gl.texParameteri(result._glTarget, device.gl.TEXTURE_MIN_FILTER, device.gl.LINEAR_MIPMAP_LINEAR);
+        result._mipmaps = true;
+
+        return result;
+    }
+
+    /**
+     * @private
+     * @function
+     * @name generateAmbient
+     * @description Generate a prefiltered ambient texture for use in IBL.
+     * @param {Texture} source - The source lighting texture, as returned from generateLightingSource.
+     * @param {object} [options] - The optional options object.
+     * @param {number} [options.numSamples] - Number of samples to use. Default is 2048.
+     * @returns {Texture} The prefiltered cubemap
+     */
+    static generateAmbient(source, options) {
+        const device = source.device;
+        const format = lightingPixelFormat(device);
+        const result = createCubemap(device, 16, format);
+
+        reprojectTexture(source, result, {
+            numSamples: options?.numSamples || 2048,
+            distribution: 'lambert'
+        });
+
+        const levels = calcLevels(result.width, result.height);
+        for (let i = 1; i < levels; ++i) {
+            const level = createCubemap(device, Math.max(1, result.width >> i), format);
+            reprojectTexture(source, level, {
+                numSamples: options?.numSamples || 2048,
+                distribution: 'lambert'
+            });
+            copyMipmap(result, level);
+            level.destroy();
+        }
+
+        // set filtering while bypassing engine
+        device.setTexture(result, 0);
+        device.gl.texParameteri(result._glTarget, device.gl.TEXTURE_MAG_FILTER, device.gl.LINEAR);
```

Comment:
do these get restored to previous values?

Reply:
Not sure what you mean. The texture shouldn't be touched by the engine again since we're not invalidating texture flags. It's the reason I set the internal`._mipmaps` member here instead of `.mipmaps`.

We actually don't want the engine to re-update this texture at all. If it did, it would call gl.generateMipmaps() since there is no CPU-side level data. Thereby killing all the mipmaps we just generated!

Reply:
ah ok, all good .. I assumed it changes some state on the device, but it only changes the texture.